R/clusterSystems.R

```diff
@@ -0,0 +1,146 @@
+#'
+#' @return A named list of length 4. Elements of the list include:
+#'
+#' \code{$Clusters} A named vector of integers from 1 to k where k is the number of clusters. Names of the elements represent the genes belonging to that cluster.
+#'
+#' \code{$Best_matches} A named vector of length k representing the cluster with the best set overlap for each input system.
+#'
+#' \code{$Jaccard_indexes} The Jaccard indexes measuring the similarity between each system and the cluster which is its best match.
+#'
+#' \code{$P_values} P values representing the probability of having a Jaccard index greater than those observed by choosing a random cluster of the same size from the set of input genes.
+#'
+#' @author \href{https://orcid.org/0000-0001-5724-2252}{Rachel Silverstein} (aut)
+#'
+#' @seealso For examples of distance functions that can be used to create custom distance matrices, see \code{\link{expr_dist}} \code{\link{jaccard_dist}} \code{\link{tf_dist}}.
+#'
+#' @examples
+#' systemDB_name <- "SysDB"
+#' myDB <- fetchData(systemDB_name)
+#' rootSysIDs <- SyDBgetRootSysIDs(myDB)
+#' sys_names <- names(rootSysIDs)
+#' systems <- SyDBgetSysSymbols(myDB, sys_names)
+#'
+#' # Cluster all of the systems in the database according to
+#' # the sum of transcription factor distance and Jaccard network distance:
+#' clusterSystems(systems,
+#'                distances = c("transcription_factor", "network_jaccard"),
+#'                combineMatrices = 'sum')
+#'
+#' # Cluster all of the systems using expression profile distance
+#' # but format it like a custom distance function
+#' GEO <- fetchData("GEOprofiles")
+#' clusterSystems(systems,
+#'                distances = NULL,
+#'                customDistanceFn = list(expr_dist),
+#'                dataSources = list(GEO),
+#'                combineMatrices = NULL)
+#'
+#'
+#'
+#' @import cluster
+#' @import VennDiagram
+#' @import igraph
+#'
+#' @export
+
+
+clusterSystems <- function(systems,
+                           distances,
+                           customDistanceFn = NULL,
+                           dataSources = NULL,
+                           combineMatrices,
+                           printVennDiagrams = TRUE) {
+
+
+  # make a single vector containing all of the genes from all of the systems together
+  all_genes <- unique(unlist(systems))
+
+  distanceMatrices <- make_distance_matrices_list(distances, all_genes)
+
+  # compute any distance matrices using custom similarity functions and data sources
+  if (!is.null(customDistanceFn)) {
+    for (i in seq_along(customDistanceFn)) {
+      data <- dataSources[[i]]
+      fn <- customDistanceFn[[i]]
+      matrix <- make_matrix(all_genes, fn, data)
+      matrix <- list(matrix)
+      names(matrix) <- paste(c("custom_function_", i), collapse = "")
+      distanceMatrices <- append(distanceMatrices, matrix)
```

Comment:
`c()` is more obvious here, we use `append()` basically only when the parameter `after` is needed.